File: src/Test/ListOpsFFI.js
export const runListOpsFFI = function(limit) {
  let n = Math.floor(limit);
  return foldl(function(acc) {
    return function(x) {
      return acc + x;
    };
  })(0)(filterEvens(range(1)(n)));
};

function range(start) {
  return function(end) {
    function go(curr, acc) {
      if (curr < start) return acc;
      return go(curr - 1, { type: "Cons", value0: curr, value1: acc });
    }
    return go(end, { type: "Nil" });
  };
}

function filterEvens(lst) {
  function go(list, acc) {
    if (list.type === "Nil") return acc;
    let x = list.value0;
    let xs = list.value1;
    if (x % 2 === 0) {
      return go(xs, { type: "Cons", value0: x, value1: acc });
    } else {
      return go(xs, acc);
    }
  }
  return go(lst, { type: "Nil" });
}

function foldl(f) {
  return function(acc) {
    return function(lst) {
      function go(list, a) {
        if (list.type === "Nil") return a;
        return go(list.value1, f(a)(list.value0));
      }
      return go(lst, acc);
    };
  };
}

File: src/Test/ListOpsFFI.go
package Test_ListOpsFFI

type ListOpsList[A any] interface {
	isList()
}
type ListOpsNil[A any] struct{}

func (ListOpsNil[A]) isList() {}

type ListOpsCons[A any] struct {
	value0 A
	value1 ListOpsList[A]
}

func (ListOpsCons[A]) isList() {}

func rangeListOps(start int, end int) ListOpsList[int] {
	var goFunc func(int, ListOpsList[int]) ListOpsList[int]
	goFunc = func(curr int, acc ListOpsList[int]) ListOpsList[int] {
		if curr < start {
			return acc
		}
		return goFunc(curr-1, ListOpsCons[int]{value0: curr, value1: acc})
	}
	return goFunc(end, ListOpsNil[int]{})
}

func filterEvens(lst ListOpsList[int]) ListOpsList[int] {
	var goFunc func(ListOpsList[int], ListOpsList[int]) ListOpsList[int]
	goFunc = func(list ListOpsList[int], acc ListOpsList[int]) ListOpsList[int] {
		switch l := list.(type) {
		case ListOpsNil[int]:
			return acc
		case ListOpsCons[int]:
			x := l.value0
			xs := l.value1
			if x%2 == 0 {
				return goFunc(xs, ListOpsCons[int]{value0: x, value1: acc})
			} else {
				return goFunc(xs, acc)
			}
		}
		return ListOpsNil[int]{}
	}
	return goFunc(lst, ListOpsNil[int]{})
}

func foldl[A, B any](f func(B) func(A) B, acc B, lst ListOpsList[A]) B {
	var goFunc func(ListOpsList[A], B) B
	goFunc = func(list ListOpsList[A], a B) B {
		switch l := list.(type) {
		case ListOpsNil[A]:
			return a
		case ListOpsCons[A]:
			return goFunc(l.value1, f(a)(l.value0))
		}
		return a
	}
	return goFunc(lst, acc)
}

func RunListOpsFFI(limit int) int {
	n := int(limit)
	res := foldl(func(acc int) func(int) int {
		return func(x int) int {
			return acc + x
		}
	}, 0, filterEvens(rangeListOps(1, n)))
	return (res)
}
